src/table.rs

```diff
@@ -273,27 +273,69 @@ impl FatTrait for Fat16 {
 impl FatTrait for Fat32 {
     fn get_raw<T: ReadSeek>(fat: &mut T, cluster: u32) -> io::Result<u32> {
         fat.seek(io::SeekFrom::Start((cluster * 4) as u64))?;
-        Ok(fat.read_u32::<LittleEndian>()? & 0x0FFFFFFF)
+        Ok(fat.read_u32::<LittleEndian>()?)
     }
 
     fn get<T: ReadSeek>(fat: &mut T, cluster: u32) -> io::Result<FatValue> {
-        let val = Self::get_raw(fat, cluster)?;
+        let val = Self::get_raw(fat, cluster)? & 0x0FFFFFFF;
         Ok(match val {
+            0 if cluster == 0x0FFFFFF7 => {
+                warn!("cluster number 0x0FFFFFF7 is a special value in FAT to indicate a BAD_CLUSTER; it should never be seen as free");
+                FatValue::Bad // avoid accidental use or allocation into a FAT chain
+            },
+            0 if cluster >= 0x0FFFFFF8 && cluster <= 0x0FFFFFFF => {
+                warn!("cluster number {} is a special value in FAT to indicate end-of-chain; it should never be seen as free", cluster);
+                FatValue::Bad // avoid accidental use or allocation into a FAT chain
+            },
             0 => FatValue::Free,
             0x0FFFFFF7 => FatValue::Bad,
             0x0FFFFFF8...0x0FFFFFFF => FatValue::EndOfChain,
+            n if cluster == 0x0FFFFFF7 => {
+                warn!("cluster number 0x0FFFFFF7 is special value in FAT to indicate a BAD_CLUSTER and thus should never be part of a FAT chain, but indicates next cluster is {}", n);
+                FatValue::Bad // avoid accidental use or allocation into a FAT chain
+            },
+            n if cluster >= 0x0FFFFFF8 && cluster <= 0x0FFFFFFF => {
+                warn!("cluster number {} is a special value in FAT to indicate end-of-chain and thus should never be part of a FAT chain, but indicates next cluster is {}", cluster, n);
+                FatValue::Bad // avoid accidental use or allocation into a FAT chain
+            },
             n => FatValue::Data(n as u32),
         })
     }
 
     fn set<T: ReadWriteSeek>(fat: &mut T, cluster: u32, value: FatValue) -> io::Result<()> {
+        let old_reserved_bits = Self::get_raw(fat, cluster)? & 0xF0000000;
         fat.seek(io::SeekFrom::Start((cluster * 4) as u64))?;
+
+        match value {
+            FatValue::Free if cluster == 0x0FFFFFF7 => {
+                let msg = "cluster number 0x0FFFFFF7 is a special value in FAT to indicate a BAD_CLUSTER; it should never be marked as free";
+                let custom_error = Error::new(ErrorKind::Other, msg);
+                return Err(custom_error);
```

Comment:
Creating error could be a one/two liner like in fs.rs. Why making additional variables here.

Reply:
changed in latest commits due to other requested changes.  Now needs to format!().